include/bout/region.hxx

```diff
@@ -0,0 +1,251 @@
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * BOUT++ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with BOUT++.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **************************************************************************/
+
+#ifndef __REGION_H__
+#define __REGION_H__
+
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "bout/openmpwrap.hxx"
+
+/*
+ * The MAXREGIONBLOCKSIZE value can be tuned to try to optimise
+ * performance on specific hardware. It determines what the largest
+ * contiguous block size can be. As we hope the compiler will vectorise
+ * the access to these contiguous blocks, the optimal MAXREGIONBLOCKSIZE
+ * is likely related to the vector size etc.
+ */
+#ifndef MAXREGIONBLOCKSIZE
+#define MAXREGIONBLOCKSIZE 64
+#endif
+#if MAXREGIONBLOCKSIZE <= 0
+#error "In region.hxx must set MAXREGIONBLOCKSIZE > 0"
+#endif
+
+// The following loops vectorise well well index is type int, needs testing
+// with current approach where index is ind2D or ind3D
+#define BLOCK_REGION_LOOP_SERIAL(region, index, ...)                                     \
+  {                                                                                      \
+    const auto blocks = region.getBlocks();                                              \
+    for (auto block = blocks.begin(); block < blocks.end(); ++block) {                   \
+      for (auto index = block->first; index <= block->second; ++index) {                 \
+        __VA_ARGS__                                                                      \
+      }                                                                                  \
+    }                                                                                    \
+  }
+
+#define BLOCK_REGION_LOOP(region, index, ...)                                            \
+  {                                                                                      \
+    const auto blocks = region.getBlocks();                                              \
+  BOUT_OMP(parallel for)                                                                 \
+    for (auto block = blocks.begin(); block < blocks.end(); ++block) {                   \
+      for (auto index = block->first; index <= block->second; ++index) {                 \
+        __VA_ARGS__                                                                      \
+      }                                                                                  \
+    }                                                                                    \
+  }
+
+class specificInd {
```

Comment:
Can we make this `SpecificInd` (capital first letter) to be consistent with other classes.